### PR TITLE
feat(openfeature): server-side EVP flagevaluation emission

### PR DESCRIPTION
**Pull request by leoromanovsky**
## Motivation

Go server-side flag evaluations need an EVP `flagevaluation` emission path alongside the existing OTel `feature_flag.evaluations` path.

The backend contract is owned by the `dd-source` `flageval-worker`: batched EVP event items accept only schema-visible fields, and OpenFeature `reason` is not an EVP event field. `targeting_rule.key` is real targeting-rule metadata, so the Go path omits `targeting_rule` until that metadata is available.

## Changes

| Area | Implementation |
| --- | --- |
| Hook path | OpenFeature `Finally` hook records EVP evaluations; the existing OTel hook path remains separate. |
| Hot path | Evaluation work is bounded to scalar extraction, flattened/pruned context snapshotting, and a non-blocking queue offer. |
| Aggregation | Background writer aggregates into full and degraded tiers, then counts drops when bounded queues or maps overflow. |
| Bucket dimensions | Full tier keys by flag, variant, allocation, runtime default, error message, targeting key, and pruned context. Degraded tier keeps flag, variant, allocation, runtime default, and error message. |
| Payload contract | SDK-owned tests assert the emitted EVP field set without copying the `dd-source` worker schema or adding a JSON Schema dependency. |
| EVP transport | Exposures and flag evaluations share one EVP client and HTTP transport; endpoints and payload shapes remain separate. |
| Benchmarks | Baseline-backed OpenFeature evaluation benchmarks from #4902 remain in `microbenchmarks-3`; on this branch the default client/provider benchmark exercises the EVP-enabled path. The focused `BenchmarkFlagEvaluation*` helpers remain runnable locally for overhead analysis, but are not listed in the benchmark gate until matching baseline names exist on `main`. |

```mermaid
flowchart LR
    eval["OpenFeature evaluation"] --> provider["DatadogProvider<br/>UFC result"]
    provider --> hooks["Finally hooks<br/>OTel + EVP"]
    hooks --> writer["EVP writer<br/>bounded queue + aggregation"]
    writer --> intake["Agent EVP proxy<br/>dd-source flageval-worker"]
```

## Decisions

- The queue stores pruned evaluation snapshots, not raw OpenFeature evaluation context.
- Full-tier context identity is a comparable canonical context string, not a hash digest.
- Degraded aggregation drops targeting key and context for cardinality control while retaining schema-visible EVP dimensions.
- Queue overflow and degraded overflow are observable counted drops; evaluation is never blocked on EVP telemetry.
- Shutdown coordinates enqueue/drain/flush so the writer has one final flush path.
- `runtime_default_used` comes from typed conversion/default behavior.
- The worker schema stays in `dd-source`; the Go SDK only tests the fields it emits.

## Validation Evidence

Local validation on June 20, 2026:

- SDK commit: `f2a6df08644a15aecaead0f68ca1e646dac276b3`
- system-tests commit: `95edcc8de242e0c70ddce87b6fab8a76e585f1e3`
- Weblog: `golang` / `net-http`
- Runtime context: Agent `7.80.1`; library `golang@2.10.0-dev`; Linux `aarch64`

Commands:

```bash
SYSTEM_TEST_BUILD_TIMEOUT=1200 ./build.sh --library golang --weblog-variant net-http
TEST_LIBRARY=golang WEBLOG_VARIANT=net-http ./run.sh +l golang FEATURE_FLAGGING_AND_EXPERIMENTATION -k "test_flag_eval_evp"
```

Result: `8 passed, 2627 deselected in 67.22s`.

Repo-local validation on June 20, 2026:

- `go test -race ./openfeature` - pass
- `go test ./openfeature -run '^(TestEndToEnd|TestFlagEvaluation|TestEvaluateStampsEvalTimeMetadata)' -count=1 -v` - pass
- `go test ./openfeature -run '^$' -bench='^BenchmarkFlagEvaluation' -benchtime=100ms -count=1` - pass
- `git diff --check` - pass

Dogfood validation:

- `SDK=go APP_URL=http://localhost:8081 WAIT_SECONDS=40 scripts/validate-server-flagevaluations.sh` with `DD_TRACE_GO_PATH=/Users/leo.romanovsky/gsd-workspaces/flag-evaluations-cross-sdk/dd-trace-go` - pass (`EVP flagevaluations OK: count=1`, `OTel non-regression OK: otel_count=1`, `service=ffe-dogfooding-go`)
- Direct backend query confirmed `flagevaluation` rows for Go targeting keys and `evaluation_count` aggregation. June 20 real-backend run id `1781931066` used prefix `server-evp-e2e-go-1781931066-user-`.

### CODE REVIEW

File: openfeature/flagevaluation_hook.go [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+// isRuntimeDefault returns true when the caller's supplied default value was returned.
+// Primary signal: absent variant key (flag-not-found, provider-not-ready, type-mismatch, no allocation).
+// Secondary: explicit DEFAULT or DISABLED reason (belt-and-suspenders).
+func isRuntimeDefault(details of.InterfaceEvaluationDetails) bool {
+	if details.Variant == "" {
+		return true
+	}
+	return details.Reason == of.DefaultReason || details.Reason == of.DisabledReason
```
Comment (dd-oleksii) [inline openfeature/flagevaluation_hook.go]:
```suggestion
	return details.Variant == ""
```

File: openfeature/flagevaluation.go [hunk 2]
```diff
@@ -0,0 +1,788 @@
...
+//	A collision requires two evaluations that are identical on ALL FIVE exact dimensions
+//	AND whose *different* contexts happen to hash to the same uint64. When that occurs, the
+//	second evaluation matches the existing bucket on add()'s fast path and increments its
+//	count (the `if e, ok := a.full[fullKey]; ok` branch below). Consequences:
+//
+//	  - COUNT-PRESERVING: the evaluation count is merged into the existing bucket, never
+//	    dropped. The invariant Σ(counts across all tiers) == number of add() calls still
+//	    holds. This is exactly what TestSaturationCountPreservation guards.
+//	  - NO MISATTRIBUTION: the count still belongs to the correct flag/variant/reason/
+//	    allocation — those dimensions are exact, not hashed.
```
Comment (dd-oleksii) [inline openfeature/flagevaluation.go]: major: on collision, it is going to misattribute the count to the wrong context _and_ miss the new context.

And I'm not sure if we actually achieve anything by hashing the context manually 🤷
- This doesn't reduce the amount of hashing as each context still needs to be hashed once
- Go's internal hashing function is faster than fnv-1a
- If we include the context here (potentially sanitized), the go will also handle hash collisions

So there's a good chance that this hashing is making performance worse. Unless we have profiles showing that context hashing is a major cost center (with go's implementation), we shouldn't even be thinking about this

File: openfeature/flagevaluation.go [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+// record runs on the evaluation hot path (the Finally hook). It does only cheap scalar
+// extraction plus the SDK's shallow context copy, then a non-blocking enqueue — no
+// flatten/prune/hash/aggregation happens here; the background worker does that. If the
+// queue is full the event is dropped and counted (best-effort), never blocking the
+// evaluation. Called from the Finally hook after every evaluation.
+func (w *flagEvaluationWriter) record(hookContext of.HookContext, details of.InterfaceEvaluationDetails) {
+	ev := evalEvent{
+		d:     extractEvalDetails(hookContext, details),
+		attrs: hookContext.EvaluationContext().Attributes(), // SDK returns an owned copy
+		nowMs: time.Now().UnixMilli(),
```
Comment (dd-oleksii) [inline openfeature/flagevaluation.go]: minor: can we pass evaluation time from the evaluator itself (e.g. via flag metadata). Getting a new timestamp here is slightly incorrect. (You may also find that getting current time may be surprisingly expensive — e.g. on big precomputed benchmark getting current time _once_ is ~1-2% of runtime.)

File: openfeature/flagevaluation.go [hunk 4]
```diff
@@ -0,0 +1,788 @@
...
+	}
+}
+
+// aggregate performs the deferred flatten/prune/hash and updates the aggregator. It runs
+// only on the writer's single worker goroutine.
+func (w *flagEvaluationWriter) aggregate(ev evalEvent) {
+	var contextAttrs map[string]any
+	if len(ev.attrs) > 0 {
+		flattened := flattenContext(ev.attrs)
+		contextAttrs = pruneContext(flattened)
```
Comment (dd-oleksii) [inline openfeature/flagevaluation.go]: Merging `flattenContext` and `pruneContext` into a single procedure is likely a good optimization opportunity — reducing allocation, improves data locality.

File: openfeature/provider_bench_test.go [hunk 5]
```diff
@@ -232,3 +243,190 @@ func BenchmarkConcurrentEvaluations(b *testing.B) {
...
+
+			flagKeys := boolBenchmarkFlagKeys(config)
+			attrs := makeBenchmarkAttrs(p.numFields)
+			ctx := context.Background()
+
+			b.ReportAllocs()
+			b.ResetTimer()
+
+			i := 0
+			for b.Loop() {
```
Comment (dd-oleksii) [inline openfeature/provider_bench_test.go:394]: Can we run a parallel benchmark? Sequential execution is untypical for Go

File: openfeature/flagevaluation_hook.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	// Finally runs — and record() is a non-blocking in-memory add with no network
+	// call. Gating on ctx.Done() would silently drop legitimate evaluation counts
+	// for cancelled-request evals.
+	if h.writer == nil {
+		return
+	}
+	h.writer.record(hookContext, details)
+}
+
+// isRuntimeDefault returns true when the caller's supplied default value was returned.
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_hook.go]: I'm not sure this is the canonical definition across SDKs; will follow up

File: openfeature/flagevaluation_hook_test.go [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+				t.Errorf("fallback first=%d not within hook-fire window [%d,%d]", e.firstEvaluation, before, after)
+			}
+		}
+	})
+}
+
+// TestFlagEvaluationHookFinally verifies that the Finally hook records an entry for
+// success, error-reason, and provider-not-ready paths.
+//
+// It must fail RED: the hook's Finally method panics with "not implemented".
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_hook_test.go]: what is this comment?

File: openfeature/flagevaluation_scale_test.go [hunk 8]
```diff
@@ -0,0 +1,356 @@
...
+		tc.sumCounts += e.count
+	}
+	tc.sumCounts += agg.droppedDegradedOverflow
+	return tc
+}
+
+// TestScaleDegradedCardinality2500Flags reports the LEGITIMATE degraded cardinality for the
+// realistic 2,500-flag shape and compares it to the production degradedCap (10,000). This is the
+// cap-sizing math that decides whether a 2-tier design drops legitimate counts.
+func TestScaleDegradedCardinality2500Flags(t *testing.T) {
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_scale_test.go:128]: what is being asserted here?

File: openfeature/flagevaluation_scale_test.go [hunk 9]
```diff
@@ -0,0 +1,356 @@
...
+			if tc.sumCounts != calls {
+				t.Errorf("count preservation violated: Σ=%d != calls=%d", tc.sumCounts, calls)
+			}
+
+			if tc.dropped > 0 {
+				t.Logf("  DROP TRIGGERED: %d evaluation(s) dropped (degraded saturated at %d).",
+					tc.dropped, defaultEvalDegradedCap)
+			} else {
+				t.Logf("  DROP NOT TRIGGERED at this sweep point — no legitimate count lost.")
+			}
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_scale_test.go]: these are logs; do we need an assert here?

File: openfeature/flagevaluation_scale_test.go [hunk 10]
```diff
@@ -0,0 +1,356 @@
...
+		t.Logf("  => all %d legitimate degraded buckets FIT under degradedCap %d (headroom %d); "+
+			"no legitimate count dropped at the scale target.",
+			deg, defaultEvalDegradedCap, defaultEvalDegradedCap-deg)
+		if tc.dropped != 0 {
+			t.Errorf("unexpected drops (%d) when legitimate cardinality fits under degradedCap", tc.dropped)
+		}
+	}
+
+	if tc.sumCounts != calls {
+		t.Errorf("count preservation violated: Σ=%d != calls=%d", tc.sumCounts, calls)
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_scale_test.go:253]: is this an assert?

File: openfeature/provider_bench_test.go [hunk 11]
```diff
@@ -168,9 +139,9 @@ func BenchmarkEvaluationWithVaryingFlagCounts(b *testing.B) {
 			provider := newDatadogProvider(ProviderConfig{})
 			config := createTestConfig()
 
-			// Add additional flags
+			// Add additional flags with unique monotonic keys so cardinality is exercised.
 			for i := len(config.Flags); i < count.numFlags; i++ {
-				flagKey := string(rune('a' + i))
+				flagKey := "flag-" + strconv.Itoa(i)
```
Comment (leoromanovsky) [inline openfeature/provider_bench_test.go]: what does this change?

File: openfeature/flagevaluation_test.go [hunk 12]
```diff
@@ -0,0 +1,1146 @@
...
+		globalCap:   globalCap,
+		perFlagCap:  perFlagCap,
+		degradedCap: degradedCap,
+	}
+}
+
+// TestPruneContext verifies that pruneContext applies the 256-field / 256-char limits
+// before evaluation context enters the aggregation buffer.
+//
+// It must fail RED: pruneContext panics with "not implemented".
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_test.go]: ?

File: openfeature/flagevaluation_test.go [hunk 13]
```diff
@@ -0,0 +1,1146 @@
...
+			tc.assert(t, pruneContext(tc.input))
+		})
+	}
+}
+
+// TestFlagEvaluationPayloadSchema verifies that full, degraded, and required-only events
+// marshal to JSON that omits the expected optional fields per tier while always including
+// the 5 required fields.
+//
+// It must fail RED: the aggregator.add method panics with "not implemented".
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_test.go]: ?

File: openfeature/flagevaluation_test.go [hunk 14]
```diff
@@ -0,0 +1,1146 @@
...
+		for _, optional := range []string{"reason", "targeting_key", "variant", "allocation", "targeting_rule", "error", "context", "runtime_default_used"} {
+			if _, ok := requiredOnly[optional]; ok {
+				t.Fatalf("required-only EVP event should omit %q", optional)
+			}
+		}
+	})
+}
+
+// TestAggregatorCollisionSafety verifies that two distinct inputs that would collide
+// under FNV-1a-only map keying land in SEPARATE buckets under the struct-keyed map.
```
Comment (leoromanovsky) [inline openfeature/flagevaluation_test.go]: are we using this hashing method?

File: openfeature/flagevaluation.go [hunk 15]
```diff
@@ -0,0 +1,902 @@
...
+	if err != nil {
+		return fmt.Errorf("failed to create request: %w", err)
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+	req.Header.Set(evpSubdomainHeader, evpSubdomainValue)
+
+	log.Debug("openfeature: sending flag evaluation events to %s", requestURL)
+
+	resp, err := w.httpClient.Do(req)
```
Comment (leoromanovsky) [inline openfeature/flagevaluation.go]: can any of this be in a shared evp client? we also post to evp for exposures. also can we avoid creating a connection?

File: openfeature/provider_bench_test.go (hunk 11 above)
Comment (leoromanovsky) [reply openfeature/provider_bench_test.go]: Resolved in 5e9b8b014. The benchmark now rotates through the generated flag keys, so the varying flag-count case measures configured flag cardinality instead of repeatedly evaluating one fixed key; I also clarified the comment at the setup site.

File: openfeature/flagevaluation_scale_test.go (hunk 10 above)
Comment (leoromanovsky) [reply openfeature/flagevaluation_scale_test.go:253]: Resolved in 5e9b8b014. This is now an assertion: the worst-case degraded-tier path fails if counts are not preserved. I also tightened the surrounding scale tests so they assert headroom, no legitimate drops, full-tier saturation, degraded overflow, and counted hot-flag drops instead of relying on logs.

File: openfeature/flagevaluation_scale_test.go (hunk 8 above)
Comment (leoromanovsky) [reply openfeature/flagevaluation_scale_test.go:128]: Resolved in 5e9b8b014. This test now asserts the production degraded cap has at least 2x headroom over the legitimate degraded cardinality for the 2,500-flag target shape; it fails if the cap falls below that rounded 2x threshold.

File: openfeature/provider_bench_test.go (hunk 5 above)
Comment (leoromanovsky) [reply openfeature/provider_bench_test.go:394]: Resolved. Added BenchmarkFlagEvaluationOTelPlusEVPParallel using b.RunParallel at the 2,500-flag / 500-user / 20-field profile with OTel + EVP enabled. It is included in the dedicated microbenchmarks-flagevaluation CI job.

File: openfeature/flagevaluation.go [hunk 16]
```diff
@@ -0,0 +1,854 @@
...
+		tmp = strconv.AppendInt(tmp, int64(x), 10)
+	case float64:
+		tag = ctxTagFloat64
+		tmp = strconv.AppendFloat(tmp, x, 'g', -1, 64)
+	case float32:
+		tag = ctxTagFloat32
+		tmp = strconv.AppendFloat(tmp, float64(x), 'g', -1, 32)
+	default:
+		tag = ctxTagOther
+		tmp = append(tmp, fmt.Sprintf("%v", x)...)
```
[the remark this answers is not in the document]
Comment (darccio) [reply openfeature/flagevaluation.go]: @leoromanovsky Does it make sense to you?

[the remark this answers is not in the document]
Comment (leoromanovsky) [reply openfeature/flagevaluation.go]: Yes; addressed in https://github.com/DataDog/dd-trace-go/pull/4886/commits/5b75720150bd47a4cfa1f9b8fa19ae4aab1bb6e6

File: .gitlab/benchmarks/micro/gitlab-ci.yml [hunk 17]
```diff
@@ -83,6 +83,14 @@ microbenchmarks-2:
     REPETITIONS: 10
     GROUP_ID: "group-2"
 
+microbenchmarks-flagevaluation:
```
Comment (darccio) [inline .gitlab/benchmarks/micro/gitlab-ci.yml]: Is this going to be only for `FlagEvaluation` or other `openfeature` benchmarks might be added in the future?

Also, if you want these benchmarks to be included in the performance gates, you need to add it below.

Keep in mind adding this will run a separate metal instance in our benchmark platform. I might be leaning to just renaming to `microbenchmarks-3`.


```suggestion
microbenchmarks-3:
```

File: .gitlab/benchmarks/micro/gitlab-ci.yml [hunk 18]
```diff
@@ -83,6 +83,14 @@ microbenchmarks-2:
...
+  extends: .microbenchmarks
+  variables:
+    BENCHMARKS: "BenchmarkFlagEvaluationNoop|BenchmarkFlagEvaluationOTelOnly|BenchmarkFlagEvaluationOTelPlusEVP|BenchmarkFlagEvaluationEVPRecord|BenchmarkFlagEvaluationOTelPlusEVPParallel"
+    CPUS_PER_BENCHMARK: 1
+    REPETITIONS: 10
+    GROUP_ID: "group-flagevaluation"
+
 pr-performance-gates:
   stage: microbenchmarks
   needs: [microbenchmarks-1, microbenchmarks-2]
```
Comment (darccio) [inline .gitlab/benchmarks/micro/gitlab-ci.yml:96]:
```suggestion
  needs: [microbenchmarks-1, microbenchmarks-2, , microbenchmarks-3]
```

File: .gitlab/benchmarks/micro/gitlab-ci.yml [hunk 19]
```diff
@@ -83,6 +83,14 @@ microbenchmarks-2:
     REPETITIONS: 10
     GROUP_ID: "group-2"
 
+microbenchmarks-flagevaluation:
+  extends: .microbenchmarks
+  variables:
+    BENCHMARKS: "BenchmarkFlagEvaluationNoop|BenchmarkFlagEvaluationOTelOnly|BenchmarkFlagEvaluationOTelPlusEVP|BenchmarkFlagEvaluationEVPRecord|BenchmarkFlagEvaluationOTelPlusEVPParallel"
+    CPUS_PER_BENCHMARK: 1
+    REPETITIONS: 10
+    GROUP_ID: "group-flagevaluation"
```
Comment (darccio) [inline .gitlab/benchmarks/micro/gitlab-ci.yml]:
```suggestion
    GROUP_ID: "group-3"
```

File: openfeature/provider_test.go [hunk 20]
```diff
@@ -255,8 +255,47 @@ func TestNewDatadogProvider(t *testing.T) {
...
+		{"flag not found", "nonexistent-flag", openfeature.FlattenedContext{"targetingKey": "user-123"}},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			before := time.Now().UnixMilli()
+			result := provider.BooleanEvaluation(ctx, tc.flagKey, false, tc.flatCtx)
+			after := time.Now().UnixMilli()
+
+			ts, ok := result.FlagMetadata[metadataEvalTimeKey].(int64)
```
Comment (darccio) [inline openfeature/provider_test.go]: Is it expected that any value in `FlagMetadata` can be read publicly?

File: openfeature/provider.go [hunk 21]
```diff
@@ -215,6 +247,11 @@ func (p *DatadogProvider) ShutdownWithContext(ctx context.Context) error {
 			p.exposureWriter.flush()
 			p.exposureWriter.stop()
 		}
+		// Stop the EVP flag evaluation writer (nil when killswitch disabled).
+		if p.flagEvalWriter != nil {
+			p.flagEvalWriter.flush()
```
Comment (darccio) [inline openfeature/provider.go]: This flush is redundant. `p.flagEvalWriter.stop` calls `drainAndFlush` which also flushes.

**Review by darccio — CHANGES_REQUESTED** [review]

File: openfeature/provider_test.go (hunk 20 above)
Comment (leoromanovsky) [reply openfeature/provider_test.go]: Thanks for asking: Yes, in OpenFeature, FlagMetadata is public on detail APIs.

The timestamp key is `dd.*` namespaced and low-sensitivity.

File: openfeature/provider.go (hunk 21 above)
Comment (leoromanovsky) [reply openfeature/provider.go]: Thanks for finding this. Yes, it is redundant, and I verified it with a targeted test.

File: .gitlab/benchmarks/micro/gitlab-ci.yml (hunk 17 above)
Comment (leoromanovsky) [reply .gitlab/benchmarks/micro/gitlab-ci.yml]: Yes at least a handful more.

**Review by darccio — APPROVED** [review]

File: openfeature/flagevaluation.go [hunk 22]
```diff
@@ -0,0 +1,855 @@
...
+			globalCap:   defaultEvalGlobalCap,
+			perFlagCap:  defaultEvalPerFlagCap,
+			degradedCap: defaultEvalDegradedCap,
+		},
+	}
+}
+
+// start begins the periodic flushing — called from InitWithContext(), NOT from the constructor.
+// Mirrors exposure.go's start() goroutine + panic recovery pattern.
+func (w *flagEvaluationWriter) start() {
```
Comment (typotter) [inline openfeature/flagevaluation.go:302]: I guess the dd-trace-go sdk doesn't have an evp buffer-and-flush mechanism already built-in? this was handy in android.

**Review by typotter — APPROVED** [review]
I'm no expert on metrics (but Vickie is). Overall, lgtm pending all of Oleksii's concerns being addressed

File: openfeature/flagevaluation.go [hunk 23]
```diff
@@ -0,0 +1,855 @@
...
+	of "github.com/open-feature/go-sdk/openfeature"
+)
+
+const (
+	// defaultFlagEvalFlushInterval is the flush interval for EVP flag evaluation events.
+	// Dedicated 10 s timer; separate from exposureWriter's 1 s interval.
+	defaultFlagEvalFlushInterval = 10 * time.Second
+
+	// flagEvaluationEndpoint is the EVP proxy endpoint for flag evaluation events.
+	flagEvaluationEndpoint = "/evp_proxy/v2/api/v2/flagevaluations"
```
Comment (petzel) [inline openfeature/flagevaluation.go]: the track name is `flagevaluation` (no trailing `s`), which should be used here in the path

File: openfeature/flagevaluation.go [hunk 24]
```diff
@@ -0,0 +1,855 @@
...
+	// The cascade is full → degraded → drop(counted). With no ultra-degraded backstop,
+	// degradedCap must hold the legitimate degraded cardinality at the team's >=2,500-flag
+	// scale target, or legitimate counts would be dropped. Degraded cardinality is based only
+	// on schema-visible dimensions retained by the degraded payload; OpenFeature reason is not
+	// an EVP field and is not part of bucket identity. globalCap (full-tier) is raised to
+	// 131,072 so a realistic 2,500-flag × multi-context workload keeps full-fidelity buckets
+	// before degrading rather than dropping them.
+	defaultEvalGlobalCap   = 131_072 // bounds full-tier buckets only; degraded is bounded separately
+	defaultEvalPerFlagCap  = 10_000  // bounds full-fidelity buckets per flag
+	defaultEvalDegradedCap = 32_768  // bounds degraded map; overflow is dropped(counted), no ultra tier
```
Comment (petzel) [inline openfeature/flagevaluation.go]: how did we come up with each of these numbers? would be nice if we could include the arithmetic rather than hard-coding, assuming these are based on some formula

File: openfeature/flagevaluation.go [hunk 25]
```diff
@@ -0,0 +1,855 @@
...
+	contextAttrs map[string]any
+	errorMessage string
+}
+
+// observe records one more evaluation against an existing bucket: it bumps the count and
+// widens the [firstEvaluation, lastEvaluation] window to include nowMs. Every existing-bucket
+// path across the three tiers funnels through here so the count++/min/max logic lives once.
+func (e *evaluationEntry) observe(nowMs int64) {
+	e.count++
+	if nowMs < e.firstEvaluation {
```
Comment (petzel) [inline openfeature/flagevaluation.go]: this seems like an invalid state. how could now be less than the first recorded evaluation? is there some async flows that could lead to this?

File: openfeature/flagevaluation.go [hunk 26]
```diff
@@ -0,0 +1,855 @@
...
+	// Per-flag cap not yet reached. Increment the attempt count for this flag
+	// regardless of whether we can actually create a full-tier bucket. This ensures
+	// the degraded overflow path activates correctly even when globalCap is full.
+	a.perFlagFull[d.flagKey]++
+
+	// Check globalCap before creating a new full-tier bucket.
+	if a.globalCount >= a.globalCap {
+		// Global full-tier cap full — count must not be lost. Route into the degraded tier
+		// (which drops targeting_key + context), sized to hold the legitimate degraded
+		// cardinality at the >=2,500-flag scale target. The per-flag attempt counter was
```
Comment (petzel) [inline openfeature/flagevaluation.go]: nit: use constant names rather than values like `2,500` in comments so they don't become stale/confusing if we were to change these values/targets

**Review by petzel — COMMENTED** [review]
looking great so far. 

Since the EvP intake only validates the API key (and not the structure of the payload), it would be good to verify this behavior end to end before we merge. Meaning, we should be able to show that we can query this data in the `flagevaluation` track using [pup](https://pup.static-app.us1.staging.dog/).

File: openfeature/flagevaluation.go (hunk 24 above)
Comment (leoromanovsky) [reply openfeature/flagevaluation.go]: Good idea!

**Comment by leoromanovsky** [comment]
> looking great so far.
> 
> Since the EvP intake only validates the API key (and not the structure of the payload), it would be good to verify this behavior end to end before we merge. Meaning, we should be able to show that we can query this data in the `flagevaluation` track using [pup](https://pup.static-app.us1.staging.dog/).

@petzel Certainly a good idea - here's the evidence.

_FYI removed from public repo._

File: openfeature/flagevaluation.go (hunk 25 above)
Comment (leoromanovsky) [reply openfeature/flagevaluation.go]: `nowMs` was a bad choice for the variable name; I renamed it to `evaluationMs`. One of the changes in this PR was to compute the timestamp at evaluation time and pass it through. These EVP events are accumulated async and can receive messages out of order; added a test to validate.

Screenshot 2026-06-18 at 10 22 17 PM

**Review by petzel — APPROVED** [review]

File: openfeature/flagevaluation.go (hunk 22 above)
Comment (leoromanovsky) [reply openfeature/flagevaluation.go:302]: Correct. In dd-trace-go we reuse the small EVP HTTP client and the exposure writer's start/stop/panic-recovery pattern, but there isn't a generic buffered EVP writer that fits this path. The existing exposure writer buffers/dedupes individual exposure events; flagevaluation has to do a different stage before flush: bound/flatten context before enqueue, aggregate counts into full/degraded tiers, preserve first/last evaluation timestamps, and enforce the per-flag/global/degraded caps. So this keeps a dedicated flagevaluation writer while sharing the transport. If we add another aggregation-style EVP stream later, that would be the right point to extract a common abstraction.

**Comment by leoromanovsky** [comment]
I dug into the OpenFeature benchmark failures and now have a clearer read on them.

Async EVP delivery remains the right architecture, but pre-enqueue work is still on the hot path and must remain small. I reduced avoidable timestamp/context/event work, kept non-blocking best-effort delivery with bounded drops under pressure, and preserved evaluation timestamps for correctness even though that means accepting a small allocation cost.

**Review by bm1549 — APPROVED** [review]
Config add LGTM

**Review by rachelyangdog — APPROVED** [review]
